### PR TITLE
netdev: Rework netdev_eth to layered structure

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR changes the netdev_eth module to a layered structure. The change here is that the netdev_driver_t functions of the netdev_eth struct are called first, and the functions of the hardware driver are called next.

The hardware drivers are adapted to not call the netdev_eth get and set functions anymore.

The `hwdev` struct member in the `netdev_eth_t` is only required to handle the L2 network stats and can be removed as soon as the stats are moved to a separate layer.

@miri64 Could it be that I'm also breaking the other (non-gnrc) network stacks here?

### Issues/PRs references

Depends on #9329

### CODE REVIEW

**Comment by miri64** [comment]
> @miri64 Could it be that I'm also breaking the other (non-gnrc) network stacks here?

The only stack effected by *this* PR could be `lwip` (all the other stacks don't support Ethernet to my knowledge, but I'm not sure about OpenThread. @jia200x?)

**Comment by miri64** [comment]
It definitely breaks `lwip`. It `tests/lwip` segfaults (reason at the moment unclear), which it definitely doesn't do in master.

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 1]
```diff
@@ -1236,9 +1236,9 @@ static void *_gnrc_netif_thread(void *args)
     netif->pid = sched_active_pid;
     /* setup the link-layer's message queue */
     msg_init_queue(msg_queue, _NETIF_NETAPI_MSG_QUEUE_SIZE);
+    dev->context = netif;
     /* register the event callback with the device driver */
     dev->event_callback = _event_cb;
-    dev->context = netif;
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c]: Still unrelated move?

**Comment by jia200x** [comment]
> The only stack effected by this PR could be lwip (all the other stacks don't support Ethernet to my knowledge, but I'm not sure about OpenThread. @jia200x?)

OpenThread only uses raw netdev. So, shouldn't break anything there

**Comment by bergzand** [comment]
Adapted LwIP to this PR

File: pkg/lwip/contrib/lwip.c [hunk 2]
```diff
@@ -73,7 +75,8 @@ void lwip_bootstrap(void)
 #ifdef MODULE_NETDEV_TAP
     for (unsigned i = 0; i < LWIP_NETIF_NUMOF; i++) {
         netdev_tap_setup(&netdev_taps[i], &netdev_tap_params[i]);
-        if (netif_add(&netif[i], &netdev_taps[i], lwip_netdev_init,
+        netdev_eth_add((netdev_t*)&netdev_taps[i], &eth_layer[i]);
+        if (netif_add(&netif[i], &eth_layer[i], lwip_netdev_init,
```
Comment (miri64) [inline pkg/lwip/contrib/lwip.c]: Ah... easier than I anticipated ^^

File: sys/net/gnrc/netif/gnrc_netif.c (hunk 1 above)
Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: Ping?

File: pkg/lwip/contrib/lwip.c (hunk 2 above)
Comment (bergzand) [reply pkg/lwip/contrib/lwip.c]: Yeah, but not a solution that scales nicely. If we ever start supporting more devices we might want to refactor this to something that scales.

File: sys/net/gnrc/netif/gnrc_netif.c (hunk 1 above)
Comment (bergzand) [reply sys/net/gnrc/netif/gnrc_netif.c]: Yeah, I have to check this, probably still unrelated, but I have a hunch that I switched the order intentionally

File: pkg/lwip/contrib/lwip.c (hunk 2 above)
Comment (miri64) [reply pkg/lwip/contrib/lwip.c]: Yupp, but then `lwip_netdev_init` needs some heavy refactoring anyways.

File: sys/net/gnrc/netif/gnrc_netif.c (hunk 1 above)
Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: Always document your work. Future you and your peers will thank you. ;-)

**Pull request closed without merging**